processors/ratelimiter/ratelimiterprocessor.go

```diff
@@ -0,0 +1,95 @@
+package ratelimiter
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	pb_struct "github.com/envoyproxy/go-control-plane/envoy/extensions/common/ratelimit/v3"
+	pb "github.com/envoyproxy/go-control-plane/envoy/service/ratelimit/v3"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/metadata"
+)
+
+type processor struct {
+	rateLimitServiceClient   pb.RateLimitServiceClient
+	domain                   string
+	domainSoftLimitThreshold uint32
+	logger                   *zap.Logger
+	tenantIDHeaderName       string
+}
+
+const (
+	TenantID       = "tenant_id"
+	GlobalTenantID = "global_tenant_id"
+)
+
+// ProcessTraces implements processorhelper.ProcessTracesFunc
+func (p *processor) ProcessTraces(ctx context.Context, traces ptrace.Traces) (ptrace.Traces, error) {
+	tenantId, err := p.getTenantId(ctx)
+	if err != nil {
+		return traces, nil
+	}
+	// two descriptors, one for tenant and other one for domain
+	desc := make([]*pb_struct.RateLimitDescriptor, 2)
+	desc[0] = &pb_struct.RateLimitDescriptor{
+		Entries: []*pb_struct.RateLimitDescriptor_Entry{
+			{
+				Key:   TenantID,
+				Value: tenantId,
+			},
+		},
+	}
+	desc[1] = &pb_struct.RateLimitDescriptor{
+		Entries: []*pb_struct.RateLimitDescriptor_Entry{
+			{
+				Key:   TenantID,
+				Value: GlobalTenantID,
+			},
+		},
+	}
+	spanCount := uint32(traces.SpanCount())
+	response, err := p.rateLimitServiceClient.ShouldRateLimit(
```

Comment:
Making an external call is going to slow down the HT Collector significantly. I'd prefer if this were done somewhere else or a network call inside the processor was avoided.

Reply:
This is already happening in istio-ingress. We have moved this to Hypertrace collector since we know the span count, which is not possible in Istio.

Reply:
@tim-mwangi thoughts about doing this async, similar to how we do ip-resolution? 

Downside would be we wouldn't rate limit immediately, but checking async would prevent slowdown.

Reply:
mmh i think the rate limiter service internal counters are also updated by making the call. @pavan-traceable does the rate-limiter consult any DBs? or is the data cached in memory.

Reply:
RateLimitService uses redis as backend store. https://github.com/envoyproxy/ratelimit this is the service which will be used.

Reply:
@pavan-traceable proceed with the impl but note we have to thoroughly test for performance and error scenarios when rate limit service is not responsive. Given that it's one request per traces batch, the impact of the latency may not be so bad.

Reply:
Sure @tim-mwangi